File: src/Types/Position.hs
module Types.Position where

import Data.List
import Data.Maybe
import Types.Basic

-- | A position
data Position
  = Quarterback
  | Halfback
  | Fullback
  | WideReceiver
  | TightEnd
  | LeftTackle
  | LeftGuard
  | Center
  | RightGuard
  | RightTackle
  | FreeSafety
  | StrongSafety
  | Cornerback
  | RightOutsideLinebacker
  | MiddleLinebacker
  | LeftOutsideLinebacker
  | RightDefensiveEnd
  | DefensiveTackle
  | LeftDefensiveEnd
  | Kicker
  | Punter
  | StrategyCard
  | NoPosition
  deriving (Eq)

positionDatas :: [(Position, EncodedPosition)]
positionDatas =
  [ (Quarterback, "Quarterback"),
    (Halfback, "Halfback"),
    (Fullback, "Fullback"),
    (WideReceiver, "Wide Receiver"),
    (TightEnd, "Tight End"),
    (LeftTackle, "Left Tackle"),
    (LeftGuard, "Left Guard"),
    (Center, "Center"),
    (RightGuard, "Right Guard"),
    (RightTackle, "Right Tackle"),
    (FreeSafety, "Free Safety"),
    (StrongSafety, "Strong Safety"),
    (Cornerback, "Cornerback"),
    (RightOutsideLinebacker, "Right Outside Linebacker"),
    (MiddleLinebacker, "Middle Linebacker"),
    (LeftOutsideLinebacker, "Left Outside Linebacker"),
    (RightDefensiveEnd, "Right Defensive End"),
    (DefensiveTackle, "Defensive Tackle"),
    (LeftDefensiveEnd, "Left Defensive End"),
    (Kicker, "Kicker"),
    (Punter, "Punter"),
    (StrategyCard, "Strategy Card"),
    (NoPosition, "")
  ]

instance Enum Position where
  fromEnum p =
    fromMaybe
      (error $ "No position found for " ++ show p)
      (findIndex ((== p) . fst) positionDatas)
  toEnum = fst . (positionDatas !!)

instance Show Position where
  show p =
    fromMaybe
      (error "No show found for Position")
      (lookup p positionDatas)

instance Ord Position where
  compare pos1 pos2 =
    compare
      (fromEnum pos1)
      (fromEnum pos2)

readToPositionData :: EncodedPosition -> Position
readToPositionData s =
  fst $
    fromMaybe
      (error $ "No read instance for " ++ s)
      (find ((== s) . snd) positionDatas)

File: src/position.py
from enum import Enum


class Position(Enum):
    QUARTERBACK = "Quarterback"
    HALFBACK = "Halfback"
    FULLBACK = "Fullback"
    WIDE_RECEIVER = "Wide Receiver"
    TIGHT_END = "Tight End"
    LEFT_TACKLE = "Left Tackle"
    LEFT_GUARD = "Left Guard"
    CENTER = "Center"
    RIGHT_GUARD = "Right Guard"
    RIGHT_TACKLE = "Right Tackle"
    FREE_SAFETY = "Free Safety"
    STRONG_SAFETY = "Strong Safety"
    CORNERBACK = "Cornerback"
    RIGHT_OUTSIDE_LINEBACKER = "Right Outside Linebacker"
    MIDDLE_LINEBACKER = "Middle Linebacker"
    LEFT_OUTSIDE_LINEBACKER = "Left Outside Linebacker"
    RIGHT_DEFENSIVE_END = "Right Defensive End"
    DEFENSIVE_TACKLE = "Defensive Tackle"
    LEFT_DEFENSIVE_END = "Left Defensive End"
    KICKER = "Kicker"
    PUNTER = "Punter"

    @staticmethod
    def offense() -> list["__class__"]:
        return [
            Position.QUARTERBACK,
            Position.HALFBACK,
            Position.FULLBACK,
            Position.WIDE_RECEIVER,
            Position.TIGHT_END,
            Position.LEFT_TACKLE,
            Position.LEFT_GUARD,
            Position.CENTER,
            Position.RIGHT_GUARD,
            Position.RIGHT_TACKLE,
        ]

    @staticmethod
    def defense() -> list["__class__"]:
        return [
            Position.FREE_SAFETY,
            Position.STRONG_SAFETY,
            Position.CORNERBACK,
            Position.RIGHT_OUTSIDE_LINEBACKER,
            Position.MIDDLE_LINEBACKER,
            Position.LEFT_OUTSIDE_LINEBACKER,
            Position.RIGHT_DEFENSIVE_END,
            Position.DEFENSIVE_TACKLE,
            Position.LEFT_DEFENSIVE_END,
        ]

    @staticmethod
    def specialTeams() -> list["__class__"]:
        return [Position.KICKER, Position.PUNTER]
